### PR TITLE
Run basic scheduler destructor inside executor context

### LINKED ISSUE #3963: `#[tokio::test]` drops future outside of async runtime
**Version**
1.8.1 + task abort fix

**Platform**
All

**Description**
`#[tokio::test]` starts a runtime and runs the future returned by the annotated function within that runtime. However, it drops the future *outside* of the runtime.

Some futures call `tokio::spawn` when dropped (in my case, the `PoolConnection` type from `sqlx`). This causes the thread to panic:

<details>
<summary>Full backtrace</summary>

```
thread 'tests::it_can_use_registry' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\sqlx-core-0.5.5\src\pool\connection.rs:74:13
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0\/library\std\src\panicking.rs:493
   1: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0\/library\core\src\panicking.rs:92
   2: core::option::expect_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0\/library\core\src\option.rs:1292
   3: core::option::Option<tokio::runtime::spawner::Spawner>::expect<tokio::runtime::spawner::Spawner>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\option.rs:349
   4: tokio::task::spawn::spawn<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\task\spawn.rs:132
   5: sqlx_core::pool::connection::{{impl}}::drop<sqlx_core::postgres::database::Postgres>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\sqlx-core-0.5.5\src\pool\connection.rs:74
   6: core::ptr::drop_in_place<sqlx_core::pool::connection::PoolConnection<sqlx_core::postgres::database::Postgres>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
   7: core::ptr::drop_in_place<generator-0>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:264
   8: core::ptr::drop_in_place<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
   9: core::ptr::drop_in_place<tokio::runtime::task::core::Stage<core::future::from_generator::GenFuture<generator-0>>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
  10: tokio::runtime::task::core::{{impl}}::set_stage::{{closure}}<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\core.rs:288
  11: tokio::loom::std::unsafe_cell::UnsafeCell<tokio::runtime::task::core::Stage<core::future::from_generator::GenFuture<generator-0>>>::with_mut<tokio::runtime::task::core::Stage<core::future::from_generator::GenFuture<generator-0>>,tuple<>,closure-0>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\loom\std\unsafe_cell.rs:14
  12: tokio::runtime::task::core::CoreStage<core::future::from_generator::GenFuture<generator-0>>::set_stage<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\core.rs:288
  13: tokio::runtime::task::core::CoreStage<core::future::from_generator::GenFuture<generator-0>>::drop_future_or_output<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\core.rs:254
  14: tokio::runtime::task::harness::cancel_task::{{closure}}<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\harness.rs:387
  15: core::ops::function::FnOnce::call_once<closure-0,tuple<>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ops\function.rs:227
  16: std::panic::{{impl}}::call_once<tuple<>,closure-0>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panic.rs:344
  17: std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure-0>,tuple<>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:379
  18: sqlx_core::postgres::connection::executor::{{impl}}::fetch_optional::{{closure}}<sqlx_core::query::Query<sqlx_core::postgres::database::Postgres, sqlx_core::postgres::arguments::PgArguments>>
  19: std::panicking::try<tuple<>,std::panic::AssertUnwindSafe<closure-0>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:343
  20: std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure-0>,tuple<>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panic.rs:431
  21: tokio::runtime::task::harness::cancel_task<core::future::from_generator::GenFuture<generator-0>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\harness.rs:386
  22: tokio::runtime::task::harness::Harness<core::future::from_generator::GenFuture<generator-0>, alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>::shutdown<core::future::from_generator::GenFuture<generator-0>,alloc::sync::Arc<tokio::runtime::basic_s
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\harness.rs:163
  23: tokio::runtime::task::raw::shutdown<core::future::from_generator::GenFuture<generator-0>,alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\raw.rs:130
  24: tokio::runtime::task::raw::RawTask::shutdown
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\raw.rs:90
  25: tokio::runtime::task::Task<alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>::shutdown<alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\mod.rs:164
  26: tokio::runtime::task::Notified<alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>::shutdown<alloc::sync::Arc<tokio::runtime::basic_scheduler::Shared>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\task\mod.rs:177
  27: tokio::runtime::basic_scheduler::{{impl}}::drop::{{closure}}<tokio::runtime::driver::Driver>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\basic_scheduler.rs:307
  28: tokio::runtime::basic_scheduler::enter::{{closure}}<closure-0,tuple<>,tokio::runtime::driver::Driver>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\basic_scheduler.rs:279
  29: tokio::macros::scoped_tls::ScopedKey<tokio::runtime::basic_scheduler::Context>::set<tokio::runtime::basic_scheduler::Context,closure-0,tuple<>>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\macros\scoped_tls.rs:61
  30: tokio::runtime::basic_scheduler::enter<closure-0,tuple<>,tokio::runtime::driver::Driver>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\basic_scheduler.rs:279
  31: tokio::runtime::basic_scheduler::{{impl}}::drop<tokio::runtime::driver::Driver>
             at C:\Users\diggs\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.4.0\src\runtime\basic_scheduler.rs:293
  32: core::ptr::drop_in_place<tokio::runtime::basic_scheduler::BasicScheduler<tokio::runtime::driver::Driver>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
  33: core::ptr::drop_in_place<tokio::runtime::Kind>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
  34: core::ptr::drop_in_place<tokio::runtime::Runtime>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ptr\mod.rs:179
  35: sqlxmq::tests::it_can_use_registry
             at .\src\lib.rs:500
  36: sqlxmq::tests::it_can_use_registry::{{closure}}
             at .\src\lib.rs:500
  37: core::ops::function::FnOnce::call_once<closure-0,tuple<>>
             at C:\Users\diggs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ops\function.rs:227
  38: core::ops::function::FnOnce::call_once
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0\library\core\src\ops\function.rs:227
```
</details>

Furthermore, despite this panic the test will pass, which means unless you pass `--nocapture` to `cargo test`, you won't be aware of the problem.

Code to reproduce (can be pasted into the `task_abort.rs` test as a regression test):
```rust

/// Checks that aborting a task whose destructor panics has the expected result.
#[test]
fn test_task_that_spawns_task_on_drop() {
    let (tx, rx) = mpsc::channel();

    {
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_time()
            .build()
            .unwrap();

        rt.block_on(async move {
            let _handle = tokio::spawn(async move {
                // Make sure the Arc is moved into the task
                let _spawn_dropped = SpawnOnDrop(tx);
                println!("task started");
                tokio::time::sleep(std::time::Duration::new(1, 0)).await
            });

            // wait for task to sleep.
            tokio::time::sleep(std::time::Duration::from_millis(10)).await;
        });
    }

    // Check that spawning the task did not panic
    assert!(rx.recv().unwrap());
}
```

### PR DESCRIPTION
**Pull request by Diggsey**
Fixes #3963

### CODE REVIEW

File: tokio/src/runtime/handle.rs [hunk 1]
```diff
@@ -301,7 +301,7 @@ impl Handle {
             .expect("failed to park thread")
     }
 
-    pub(crate) fn shutdown(mut self) {
+    pub(crate) fn shutdown(&mut self) {
         self.spawner.shutdown();
     }
```
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:306]: Why this?

Comment (Diggsey) [reply tokio/src/runtime/handle.rs:306]: Adding a destructor to `Runtime` means we can no longer move out of the `handle` field, which would have made it impossible to call `shutdown()` on it. This was the simplest fix.

Comment (Diggsey) [reply tokio/src/runtime/handle.rs:306]: (See the implementation of `shutdown_timeout`)

**Comment by Darksonn** [comment]
The CI failed again after restarting. I suspect there is a double panic somewhere. Try running them locally with `--nocapture` to see if you get any output before it aborts.

**Comment by Diggsey** [comment]
Output is as expected when I run it locally: each of the panic tests panics once.

**Pull request closed without merging**